sdk/src/main/java/com/deploygate/sdk/DeployGate.java

```diff
@@ -1390,4 +1402,192 @@ public static String getDistributionUserName() {
+        if (sInstance != null) {
+            return sInstance.putRuntimeExtraValueInternal(key, value);
+        }
+        return false;
+    }
+
+    public static boolean putRuntimeExtraValue(String key, boolean value) {
+        if (sInstance != null) {
+            return sInstance.putRuntimeExtraValueInternal(key, value);
+        }
+        return false;
+    }
+
+    public static void removeRuntimeExtraValue(String key) {
+        if (sInstance != null) {
+            sInstance.removeRuntimeExtraValueInternal(key);
+        }
+    }
+
+    public static void removeAllRuntimeExtraValues() {
+        if (sInstance != null) {
+            sInstance.removeAllRuntimeExtraValuesInternal();
+        }
+    }
+
+    private boolean putBuildEnvironmentValueInternal(String key, Object value) {
+        if (mBuildEnvironments.size() >= MAX_BUILD_ENVIRONMENTS_SIZE && !mBuildEnvironments.containsKey(key)) {
+            Log.w(TAG, "BuildEnvironment already full. Ignored: " + key);
+            return false;
+        }
+
+        if (!isValidKey(key)) {
+            return false;
+        }
+
+        if (!isValidValue(value)) {
+            return false;
+        }
+
+        mBuildEnvironments.put(key, value);
+        return true;
+    }
+
+    private void removeBuildEnvironmentValueInternal(String key) {
+        mBuildEnvironments.remove(key);
+    }
+
+    private void removeAllBuildEnvironmentValuesInternal() {
+        mBuildEnvironments.clear();
+    }
+
+    private boolean putRuntimeExtraValueInternal(String key, Object value) {
+        if (mRuntimeExtras.size() >= MAX_RUNTIME_EXTRAS_SIZE && !mRuntimeExtras.containsKey(key)) {
+            Log.w(TAG, "RuntimeExtra already full. Ignored: " + key);
+            return false;
+        }
+
+        if (!isValidKey(key)) {
+            return false;
+        }
+
+        if (!isValidValue(value)) {
+            return false;
+        }
+
+        mRuntimeExtras.put(key, value);
+        return true;
+    }
+
+    private void removeRuntimeExtraValueInternal(String key) {
+        mRuntimeExtras.remove(key);
+    }
+
+    private void removeAllRuntimeExtraValuesInternal() {
+        mRuntimeExtras.clear();
+    }
+
+    private boolean isValidKey(String key) {
+        if (key == null || key.equals("true") || key.equals("false") || key.equals("null")) {
+            Log.w(TAG, "Not allowed key: " + key);
+            return false;
+        }
+
+        if (key.length() < MIN_USER_INPUT_KEY_LENGTH || key.length() > MAX_USER_INPUT_KEY_LENGTH || !USER_INPUT_KEY_PATTERN.matcher(key).matches()) {
+            Log.w(TAG, "Invalid key: " + key);
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean isValidValue(Object value) {
+        if (value == null) {
+            Log.w(TAG, "Value is null");
+            return false;
+        }
+
+        if (value instanceof String && ((String) value).length() > MAX_USER_INPUT_VALUE_LENGTH) {
```

Comment:
Please check `Object` types. String, Number, Boolean are only allowed, right?

Reply:
> String, Number, Boolean are only allowed, right?

Yes.
I forgot to check the value is instance of Number or Boolean after this line.